### PR TITLE
Ozan/merkletree

### PR DESCRIPTION
**Pull request by onacitarhan17**

### CODE REVIEW

File: src/test/java/unittest/fixtures/MerkleTreeFixture.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * Creates a new randomly looking MerkleTree.
+ */
+public class MerkleTreeFixture {
+  /**
+   * Creates a new skip list with n random elements.
+   *
+   * @param n number of elements to create
+   * @return a new skip list with n random elements
+   */
+  public static MerkleTree createSkipList(int n) {
```
Comment (thep2p) [inline src/test/java/unittest/fixtures/MerkleTreeFixture.java]: Seems like there are still residues of SkipList in this PR, please check and fix.

File: src/test/java/modules/ads/MerkleTreeTest.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+import modules.ads.merkletree.Verifier;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import unittest.fixtures.EntityFixture;
+import unittest.fixtures.MerkleTreeFixture;
+
+/**
+ * Encapsulates tests for an authenticated and concurrent implementation of SkipList ADS.
+ */
+public class MerkleTreeTest {
```
Comment (thep2p) [inline src/test/java/modules/ads/MerkleTreeTest.java:24]: Please add a comment for each test briefly explaining what it evaluates.

File: src/main/java/modules/ads/merkletree/MerkleTree.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      return new AuthenticatedLightChainEntity(proof, e.type(), e);
+    }
+  }
+
+  @Override
+  public AuthenticatedEntity get(Entity e) {
+    Proof proof;
+    if (e == null) {
+      return null;
+    }
```
Comment (thep2p) [inline src/main/java/modules/ads/merkletree/MerkleTree.java]: Invalid input must trigger an `IllegalArgumentException`, the `get` should explicitly list the exceptions it throws, and those exceptions must be bubbled up to the caller, literally safely fail the main. Also, please add a test covering this test case.

File: src/main/java/modules/ads/merkletree/MerkleTree.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    lock.readLock().lock();
+    try {
+      proof = getProof(e.id());
+    } finally {
+      lock.readLock().unlock();
+    }
+    if (proof == null) {
+      return null;
+    }
```
Comment (thep2p) [inline src/main/java/modules/ads/merkletree/MerkleTree.java]: On a happy path, do we expect to have a `null` proof? if we don't, then this line steps into an illegal state, and hence must throw an `IllegalStateException`. (Also, please add test covering this test case).

File: src/main/java/modules/ads/merkletree/MerkleTree.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    return new AuthenticatedLightChainEntity(proof, e.type(), e);
+  }
+
+  private Proof getProof(Identifier id) {
+    Integer idx = leafNodesHashTable.get(hasher.computeHash(id));
```
Comment (thep2p) [inline src/main/java/modules/ads/merkletree/MerkleTree.java]: `id`s are unique, i.e., they correspond to the hash of entities they keep, hence, there is no need to hash them again, right?

File: src/main/java/modules/ads/merkletree/MerkleTree.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    if (idx == null) {
+      return null;
+    }
```
Comment (thep2p) [inline src/main/java/modules/ads/merkletree/MerkleTree.java]: Isn't it again an `IllegalArgumentException` case?

File: src/main/java/modules/ads/merkletree/MerkleTree.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    this.lock = new ReentrantReadWriteLock();
+    this.leafNodesHashTable = new HashMap<>();
+  }
+
+  @Override
+  public AuthenticatedEntity put(Entity e) {
+    lock.writeLock().lock();
+    if (e == null) {
+      lock.writeLock().unlock();
+      return null;
```
Comment (thep2p) [inline src/main/java/modules/ads/merkletree/MerkleTree.java]: Isn't it an `IllegalArgumentException` case?

File: src/main/java/modules/ads/merkletree/MerkleTree.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+  public MerkleTree() {
+    this.size = 0;
+    this.root = new MerkleNode();
+    this.leafNodes = new ArrayList<>();
+    this.lock = new ReentrantReadWriteLock();
+    this.leafNodesHashTable = new HashMap<>();
+  }
+
+  @Override
+  public AuthenticatedEntity put(Entity e) {
```
Comment (thep2p) [inline src/main/java/modules/ads/merkletree/MerkleTree.java]: A general comment about this and some subsequent methods, where unlocking the lock is happening on several disjoint execution paths. It makes the implementation daunting to keep track of lock-unlock state transitions and makes it bug-prone regarding the liveness issues. Hence, please use [`try-finally`](https://docs.oracle.com/javase/tutorial/essential/exceptions/finally.html) blocks to lock and unlock only once.

File: src/main/java/modules/ads/AuthenticatedDataStructure.java [hunk 9]
```diff
@@ -1,13 +1,12 @@
...
-import model.lightchain.Identifier;
 
 /**
  * Models AuthenticatedDataStructure (ADS) and a key-value store of entities supported with membership proofs.
  */
 public interface AuthenticatedDataStructure {
   AuthenticatedEntity put(Entity e);
 
-  AuthenticatedEntity get(Identifier id);
+  AuthenticatedEntity get(Entity e);
```
Comment (thep2p) [inline src/main/java/modules/ads/AuthenticatedDataStructure.java]: This is wrong! `get` is utilized only when we have the `id` of the `entity` and not the `entity` itself. Please abide to the original interface.

File: src/main/java/crypto/Sha3256Hasher.java [hunk 10]
```diff
@@ -29,4 +40,45 @@ public Sha3256Hash computeHash(EncodedEntity e) {
...
+
+  /**
+   * Commutative hashing of two given byte arrays.
+   *
+   * @param b1 first byte array.
+   * @param b2 second byte array.
+   * @return SHA3-256 hash object of the commutative concatenation (i.e. max(b1,b2) || min(b1,b2))
+   * of the two byte arrays.
+   */
+  public Sha3256Hash computeHash(byte[] b1, byte[] b2) {
```
Comment (thep2p) [inline src/main/java/crypto/Sha3256Hasher.java:67]: why not always `h(b1||b2)`?

File: src/test/java/modules/ads/MerkleTreeTest.java [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    AtomicInteger threadError = new AtomicInteger();
+    CountDownLatch countDownLatch = new CountDownLatch(concurrencyDegree);
+    Thread[] merkleTreeThreads = new Thread[concurrencyDegree];
+    MerkleTree merkleTree = MerkleTreeFixture.createSkipList(5);
+    for (int i = 0; i < concurrencyDegree; i++) {
+      merkleTreeThreads[i] = new Thread(() -> {
+        Entity entity = new EntityFixture();
+        try {
+          merkleTree.put(entity);
+          AuthenticatedEntity authenticatedEntity = merkleTree.get(entity);
```
Comment (thep2p) [inline src/test/java/modules/ads/MerkleTreeTest.java]: get should be done in a separate for loop (sequentially), once all `put` threads are done.

File: src/test/java/unittest/fixtures/MerkleTreeFixture.java (hunk 1 above)
Comment (onacitarhan17) [reply src/test/java/unittest/fixtures/MerkleTreeFixture.java]: fixed in https://github.com/yhassanzadeh13/lightchain-sdp/pull/32/commits/29bb463d0eb39091a9e52fee2cce2bfbf95a482b

File: src/test/java/modules/ads/MerkleTreeTest.java (hunk 2 above)
Comment (onacitarhan17) [reply src/test/java/modules/ads/MerkleTreeTest.java:24]: fixed in https://github.com/yhassanzadeh13/lightchain-sdp/pull/32/commits/df3b4332ea8c901784d87ed1d7dd494d2459d0bf

File: src/main/java/modules/ads/merkletree/MerkleTree.java (hunk 3 above)
Comment (onacitarhan17) [reply src/main/java/modules/ads/merkletree/MerkleTree.java]: fixed in https://github.com/yhassanzadeh13/lightchain-sdp/pull/32/commits/94cfe9bd0d5e163b9ae22553b839d4e6cb93dc0e

File: src/main/java/modules/ads/merkletree/MerkleTree.java (hunk 7 above)
Comment (onacitarhan17) [reply src/main/java/modules/ads/merkletree/MerkleTree.java]: fixed in https://github.com/yhassanzadeh13/lightchain-sdp/pull/32/commits/94cfe9bd0d5e163b9ae22553b839d4e6cb93dc0e

File: src/main/java/modules/ads/merkletree/MerkleTree.java (hunk 5 above)
Comment (onacitarhan17) [reply src/main/java/modules/ads/merkletree/MerkleTree.java]: fixed in https://github.com/yhassanzadeh13/lightchain-sdp/pull/32/commits/d3b1343c705ed3440ea4ed551c09c564515e0f76

File: src/main/java/modules/ads/merkletree/MerkleTree.java (hunk 8 above)
Comment (onacitarhan17) [reply src/main/java/modules/ads/merkletree/MerkleTree.java]: fixed in https://github.com/yhassanzadeh13/lightchain-sdp/pull/32/commits/ccffc171c15990d92443776cd6df8067827ce2ca

File: src/main/java/modules/ads/AuthenticatedDataStructure.java (hunk 9 above)
Comment (onacitarhan17) [reply src/main/java/modules/ads/AuthenticatedDataStructure.java]: fixed in https://github.com/yhassanzadeh13/lightchain-sdp/pull/32/commits/cc555b4babb882ce872821a6449aeafa1cdfc0f7

File: src/test/java/modules/ads/MerkleTreeTest.java (hunk 11 above)
Comment (onacitarhan17) [reply src/test/java/modules/ads/MerkleTreeTest.java]: fixed in https://github.com/yhassanzadeh13/lightchain-sdp/pull/32/commits/8cdb9a28db950d6a3d02dfea8e81c543635be578

File: src/main/java/modules/ads/merkletree/MerkleTree.java (hunk 6 above)
Comment (onacitarhan17) [reply src/main/java/modules/ads/merkletree/MerkleTree.java]: fixed in https://github.com/yhassanzadeh13/lightchain-sdp/pull/32/commits/205970f31a77d2becb838840311053949f68d4e4

File: src/main/java/modules/ads/merkletree/MerkleTree.java (hunk 4 above)
Comment (onacitarhan17) [reply src/main/java/modules/ads/merkletree/MerkleTree.java]: fixed in https://github.com/yhassanzadeh13/lightchain-sdp/pull/32/commits/205970f31a77d2becb838840311053949f68d4e4

File: src/main/java/crypto/Sha3256Hasher.java (hunk 10 above)
Comment (onacitarhan17) [reply src/main/java/crypto/Sha3256Hasher.java:67]: fixed in https://github.com/yhassanzadeh13/lightchain-sdp/pull/32/commits/d0f93abbd7cd5216c659633295de486e6f16bc1e

**Pull request merged**